### PR TITLE
Renderer: Threaded frame dumping.

### PR DESCRIPTION
**Pull request by degasus**
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/4345)

### CODE REVIEW

**Comment by lioncash** [comment]
Review status: 0 of 2 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 556 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpTAcd4bkE2UIHPKpH:-KTpTAcd4bkE2UIHPKpI:b-77ow5b) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L556)):*
> ```cpp
> void Renderer::ShutdownFrameDumping()
> {
>   if (m_framedump_thread_running.IsSet())
> ```

```cpp
if (!m_framedump_thread_running.IsSet())
  return;
```

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 585 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpTLMuXSAxw00EGs7Y:-KTpTLMuXSAxw00EGs7Z:bevfqls) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L585)):*
> ```cpp
> void Renderer::FinishFrameData()
> {
>   if (m_framedump_frame_running)
> ```

```
if (!m_framedump_frame_running)
  return;
```

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 595 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpUZrtxQYj6qXHEkVY:-KTpUZrtxQYj6qXHEkVZ:becj0h0) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L595)):*
> ```cpp
> {
>   Common::SetCurrentThreadName("FrameDumping");
>   bool AVI_dump_started = false;
> ```

`avi_dump_started`

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 645 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpTcNj97MvZolCq05d:-KTpTcNj97MvZolCq05e:barstk) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L645)):*
> ```cpp
>         {
>           SConfig::GetInstance().m_DumpFrames = false;
>           OSD::AddMessage("AVIDump Start failed");
> ```

Considering how generic this is, this is a message that should probably be done within `AVIDump::Start`

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 660 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpTosuCw0weUzXuAMO:-KTpTosv1yv6hVkKgxjp:bjmj5vz) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L660)):*
> ```cpp
>     AVI_dump_started = false;
>     AVIDump::Stop();
>     OSD::AddMessage("Stop dumping frames");
> ```

Ditto here, this should likely be within AVIDump::Stop.

This should also use "Stopped" rather than "Stop"

---

*[Source/Core/VideoCommon/RenderBase.h, line 195 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpUNHyLxiEFk0sldRp:-KTpUNHyLxiEFk0sldRq:b-pdph6o) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.h#L195)):*
> ```objc
> 
>   // framedumping
>   std::thread m_framedump_thread;
> ```

`framedump` isn't a single word.

---

*[Source/Core/VideoCommon/RenderBase.h, line 208 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpUQU5C1meCQ8aLfg0:-KTpUQU5C1meCQ8aLfg1:b-uyy4su) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.h#L208)):*
> ```objc
>     u64 ticks;
>     bool upside_down;
>   } m_framedump;
> ```

`m_frame_dump_config`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345)*

**Comment by RisingFog** [comment]
Tested savestates with this PR, does not seem to desync audio and video when dumping and loading savestates. Must be thinking of #4344

**Comment by degasus** [comment]
DC without any syncing is broken. Blocking GPU features like EFB2RAM, EFB access, real XFB, frame dumping make them worse. This PR however should improve this a bit, not make it worse. But this issue called "dual core" is also on master.

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 556 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpTAcd4bkE2UIHPKpH:-KVsyJ9ErrWQY2_JJbn_:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L556)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> if (!m_framedump_thread_running.IsSet())
>   return;
> ```
</details>
Done.

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 585 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpTLMuXSAxw00EGs7Y:-KVsyP-zEZ49xxbwYFed:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L585)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```
> if (!m_framedump_frame_running)
>   return;
> ```
</details>
Done.

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 595 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpUZrtxQYj6qXHEkVY:-KVsyXcdf4R_PgaqizVm:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L595)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `avi_dump_started`
</details>
Done.

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 645 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpTcNj97MvZolCq05d:-KVsyYwjSblrMDTaCNEf:b-ankh5h) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L645)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Considering how generic this is, this is a message that should probably be done within `AVIDump::Start`
</details>
They are moved in PR #4423  already.

---

*[Source/Core/VideoCommon/RenderBase.cpp, line 660 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpTosuCw0weUzXuAMO:-KVsyk3wjf1jhX3AinBk:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.cpp#L660)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Ditto here, this should likely be within AVIDump::Stop.
> 
> This should also use "Stopped" rather than "Stop"
</details>
Done.

---

*[Source/Core/VideoCommon/RenderBase.h, line 195 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpUNHyLxiEFk0sldRp:-KVsyyOXwuN0N6BAfIjL:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.h#L195)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `framedump` isn't a single word.
</details>
Done.

---

*[Source/Core/VideoCommon/RenderBase.h, line 208 at r1](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KTpUQU5C1meCQ8aLfg0:-KVsz18-P-rvF20wJ7x0:b-896fix) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a6d7bf937019d9febc0df54d284d4c22c8307b8d/Source/Core/VideoCommon/RenderBase.h#L208)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `m_frame_dump_config`
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-:-KVt-RdGR5XS529FCit3:b-pizli9)*

**Comment by RisingFog** [comment]
Dual Core issues are related to SyncGPU being disabled, so not a regression.

Not able to reproduce the RealXFB issue anymore.

:lgtm: 

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-:-KVuQQGi8v6zDGtAwJBn:b3u43t7)*

**Comment by degasus** [comment]
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[Source/Core/VideoCommon/RenderBase.h, line 199 at r2](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-KVy_bwwMeC80yps3nKB:-KVy_bwwMeC80yps3nKC:b-c42nu1) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c9dd560abeedcc7d04f8fc45f09e66b26ef74d99/Source/Core/VideoCommon/RenderBase.h#L199)):*
> ```objc
>   Common::Event m_frame_dump_start;
>   Common::Event m_frame_dump_done;
>   Common::Flag m_frame_dump_thread_running;
> ```

Stenzek:
Could you replace m_frame_dump_thread_running with just testing m_frame_dump_thread.joinable()?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345)*

**Comment by stenzek** [comment]
:lgtm: apart from possibly removing the m_frame_dump_thread_running (but this would require a different termination condition for the loop in the worker thread).

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4345#-:-KVyamTdvg6jZrYWCbg5:b-eft6w8)*

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 1]
```diff
@@ -654,65 +650,117 @@ bool Renderer::IsFrameDumping()
...
+    data.assign(config.data, config.data + config.stride * config.height);
+
+    // As we've done a copy now, there is no need to block the GPU thread any longer.
+    m_frame_dump_done.Set();
+
+    if (config.upside_down)
+      FlipImageData(data.data(), config.width, config.height, 4);
+
+    // Save screenshot
+    if (s_bScreenshot)
```
Comment (degasus) [inline Source/Core/VideoCommon/RenderBase.cpp]: s_bScreenshot needs to be protected by s_criticalScreenshot

**Pull request merged**